### ISSUE TITLE
How would journaling work?

### ISSUE THREAD

**Issue by FreekPaans** [mzheravin/exchange-core#44]
I wonder how you'd do it and how it would impact performance? Won't this create a bottleneck (at least per order book)?

**Comment by mzheravin** [comment]
Journalling is just a streaming compression of incoming messages.  
For CPU it is 100 times easier than doing order matching and risk evaluation, therefore no significant throughtput impact is expected. 
Latency can be a problem if doing synchronious disk writes. But, there are some workarounds. For example, some exchanges are ok with memory mapped files (though I would not recommend that); or data can be writtten to the flash backed memory of the RAID controller; or it can be confirmed RDMA write, etc.

**Comment by FreekPaans** [comment]
But won't synchronous writes be a requirement in a production environment, for auditing purposes? I guess my question is: won't writing to disk be the effective bottleneck in this system - not the matching engine?

**Comment by mzheravin** [comment]
Yes, there is a requirement for synchrnious writes. If we assume main system failure does not affect RAID controller or remote machine - the methods I have mentioned above are synchronious.
Yes, journaling will cause a latency bottleneck for lower percentiles. But not for throughtput, thanks to the batching in the disruptor.

**Comment by qingtiandalaoye** [comment]
I think mmap file is nice , would you pls talk about why dont recommend it?
